### PR TITLE
Fix a serious resource leak in PWA tests on macOS

### PR DESCRIPTION
**Pull request by mherrmann**
`browser_tests --gtest_filter='WebAppIntegration.*'` and others leaked processes and clogged up the system's `sandboxd` when the tests were run from a volume that macOS classifies as "removable". A more comprehensive root cause analysis is given [here](https://github.com/brave/devops/issues/16445#issuecomment-5023870878).

Resolves https://github.com/brave/devops/issues/16445.

The best entry point for reviewing the PR is `app_shim_launch.mm`.

I'm labeling this as `QA/No` because I feel that the affected functionality is covered well by automated tests.

### CODE REVIEW

File: chromium_src/chrome/app_shim/DEPS [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from brave_chromium_utils import inline_file
+inline_file("//chrome/app_shim/DEPS", globals(), locals())
```
Comment (goodov) [inline chromium_src/chrome/app_shim/DEPS]: this pattern shouldn't be necessary in chromium_src DEPS files, because these rules are applied automatically here. This only works in chromium_src though. In other `brave/` dirs this pattern does make sense.

File: components/constants/BUILD.gn [hunk 2]
```diff
@@ -30,6 +30,7 @@ source_set("constants") {
     "brave_constants.h",
     "brave_paths.h",
     "brave_switches.h",
+    "mac_app_mode_common.h",
```
Comment (goodov) [inline components/constants/BUILD.gn]: please don't add new files to this target.

pls put this either into `brave/app_shim` or `brave/browser/app_shim`.

File: patches/chrome-app_shim-app_shim_controller.mm.patch [hunk 3]
```diff
@@ -0,0 +1,12 @@
+diff --git a/chrome/app_shim/app_shim_controller.mm b/chrome/app_shim/app_shim_controller.mm
+index 882808b999db0655873acb9417264cce7bcdea8b..efd55de59bc24db362c816715f25f3db5ef9bb12 100644
+--- a/chrome/app_shim/app_shim_controller.mm
++++ b/chrome/app_shim/app_shim_controller.mm
+@@ -633,6 +633,7 @@ void AppShimController::SendBootstrapOnShimConnected(
+              app_mode::kIsNormalLaunch))
+           ? chrome::mojom::AppShimLaunchType::kRegisterOnly
+           : chrome::mojom::AppShimLaunchType::kNormal;
++  BRAVE_SEND_BOOTSTRAP_ON_SHIM_CONNECTED
```
Comment (goodov) [inline patches/chrome-app_shim-app_shim_controller.mm.patch]: if you feel comfortable, you can try to use plaster rewrite called `preempt_function_impl`. Look for examples in `brave/rewrite/`.

we need to try to stop relying on `#define` replacements in new code.

**Review by goodov — APPROVED** [review]
the change looks good. I don't see another way of solving this (I've asked llm and searched a bit).

ideally this should be upstreamed as the problem is real and is related to `app_shim` logic not taking into account a possible environment edge case.

until then, this is fine, but it would be nice to use plaster rewrites instead of `#define` replacements.

File: chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+#include <chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm>
+
+#undef BRAVE_LAUNCH_THE_FIRST_SHIM_THAT_WORKS_ON_FILE_THREAD
+#undef LaunchShimForTesting
+
+namespace web_app {
+
+// Replaces the upstream LaunchServices-based implementation with a direct
+// spawn.
+void LaunchShimForTesting(const base::FilePath& shim_path,  // IN-TEST
```
Comment (goodov) [inline chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm]: can we call this via preempt rewrite in the upstream function and always return? That approach would be better than a function rename via `#define`.

File: patches/chrome-browser-web_applications-os_integration-mac-app_shim_launch.mm.patch [hunk 5]
```diff
@@ -0,0 +1,12 @@
+diff --git a/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm b/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm
+index 49761f6f52c57c807c516cdb150b91ab7e5c1c91..4e28737618608af408cec0e66e604b0ec5f0fd14 100644
+--- a/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm
++++ b/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm
+@@ -159,6 +159,7 @@ void LaunchTheFirstShimThatWorksOnFileThread(
+   variations::VariationsCommandLine::GetForCurrentProcess().ApplyToCommandLine(
+       command_line);
+ 
++  BRAVE_LAUNCH_THE_FIRST_SHIM_THAT_WORKS_ON_FILE_THREAD
```
Comment (goodov) [inline patches/chrome-browser-web_applications-os_integration-mac-app_shim_launch.mm.patch]: if you can implement this as a plaster rewrite, it would be great, but I believe the only option available is a regex for this kind of patch.. I'm not a fan of those and `#define`-insert is acceptable here imo.

File: chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm (hunk 4 above)
Comment (mherrmann) [reply chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm]: 55575ac84ed853e6a5da6450ef3e8c12ac1c08b8

File: patches/chrome-browser-web_applications-os_integration-mac-app_shim_launch.mm.patch (hunk 5 above)
Comment (mherrmann) [reply patches/chrome-browser-web_applications-os_integration-mac-app_shim_launch.mm.patch]: I had to rewrite it as a plaster rewrite because your other suggestion made me create a plaster rewrite for the file, which then invalidated the patch that referenced the macro. It's a regex, but not too bad imo.

f7070c75373472d4ba1fd429160655fcd1f8afff

**Review by netzenbot — APPROVED** [review]

File: chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+  const base::FilePath executable_path =
+      base::apple::NSURLToFilePath(bundle.executableURL.absoluteURL);
+  if (executable_path.empty()) {
+    LOG(ERROR) << "Could not resolve executable in bundle: " << shim_path;
+    return base::Process();
+  }
+  base::CommandLine shim_command_line(executable_path);
+  shim_command_line.AppendArguments(command_line, /*include_program=*/false);
+  return base::LaunchProcess(shim_command_line, {});
+}
```
[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/chrome/browser/web_applications/os_integration/mac/app_shim_launch.mm:70]: 1. ThreadPool workers already scope an autorelease pool per work cycle. In
  base/task/thread_pool/worker_thread.cc:436, RunWorker() creates an apple::ScopedNSAutoreleasePool at
  the top of every iteration of its work loop, on Apple platforms. Both call paths of LaunchShimDirectly
  run on ThreadPool workers (via internals::GetShortcutIOTaskRunner(), and upstream's
  LaunchShimOnFileThread posts to the same runner), so autoreleased NSBundle/NSURL objects are drained
  after each work batch — never accumulated for the thread's lifetime. The same is true for base::Threads
  with message pumps, which drain a pool per pump cycle.
2. Upstream does exactly the same thing in the same file. The unmodified app_shim_launch.mm creates
  NSBundle bundleWithURL:, NSWorkspace.sharedWorkspace.runningApplications, NSStrings, etc. on these same
  threads with no local @autoreleasepool (e.g. line 93 of the upstream file — nearly identical code to
  LaunchShimDirectly). Not wrapping individual Foundation calls in pools is the established Chromium
  convention precisely because the threading infrastructure owns the pool.

**Review by goodov — APPROVED** [review]

**Comment by wknapik** [comment]
Can we uplift?

**Comment by mherrmann** [comment]
Even though the functionality is covered by tests, I do not know if there are edge cases that are not covered by tests and which might cause problems. I would therefore prefer to let this bake. If we wanted, we could disable the tests in CI until this lands in prod.
